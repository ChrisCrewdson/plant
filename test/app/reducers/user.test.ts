import si from 'seamless-immutable';

// @ts-ignore
const seamless = si.static;
const user = require('../../../app/reducers/user');
const { actionFunc } = require('../../../app/actions');

function checkReducer(actionName: string, state: any, payload?: object) {
  const action = actionFunc[actionName](payload);
  const actual = user(state, action);

  expect(actual).toMatchSnapshot();
}

describe('/app/reducers/user', () => {
  test('should reduce a logout request', () => {
    const state = seamless.from({});
    const payload = {};

    checkReducer('logoutRequest', state, payload);
  });

  test('should reduce a logout success', () => {
    const state = seamless.from({});
    const payload = {};

    checkReducer('logoutSuccess', state, payload);
  });

  test('should reduce a logout failure', () => {
    const state = seamless.from({});
    const payload = {};

    checkReducer('logoutFailure', state, payload);
  });

  test('should reduce a loadLocationsSuccess', () => {
    const state = seamless.from({
      _id: 'u-1',
      isLoggedIn: true,
      activeLocationId: false,
    });

    const payload = [{
      members: {
        'u-1': true,
      },
      _id: 'l-1',
    }];

    checkReducer('loadLocationsSuccess', state, payload);
  });

  test('should reduce a missing location', () => {
    const state = seamless.from({
      _id: 'u-1',
      isLoggedIn: true,
      activeLocationId: false,
    });

    const payload = [{
      members: {
        'u-2': true,
      },
      _id: 'l-1',
    }];

    checkReducer('loadLocationsSuccess', state, payload);
  });

  test('should not change state with loadLocationsSuccess when not logged in', () => {
    const state = seamless.from({
      _id: 'u-1',
      isLoggedIn: false,
      activeLocationId: false,
    });

    const payload = undefined;

    checkReducer('loadLocationsSuccess', state, payload);
  });

  test('should handle changeActiveLocationId', () => {
    const state = seamless.from({
      activeLocationId: 'l-2',
    });

    const payload = {
      _id: 'l-1',
    };

    checkReducer('changeActiveLocationId', state, payload);
  });

  test('should handle changeActiveLocationId with no payload', () => {
    const state = seamless.from({
      activeLocationId: 'l-2',
    });

    const payload = undefined;

    checkReducer('changeActiveLocationId', state, payload);
  });
});
