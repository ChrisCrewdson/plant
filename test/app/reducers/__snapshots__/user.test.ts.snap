// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/app/reducers/user should handle changeActiveLocationId 1`] = `
Object {
  "activeLocationId": "l-1",
}
`;

exports[`/app/reducers/user should handle changeActiveLocationId with no payload 1`] = `
Object {
  "activeLocationId": "l-2",
}
`;

exports[`/app/reducers/user should not change state with loadLocationsSuccess when not logged in 1`] = `
Object {
  "_id": "u-1",
  "activeLocationId": false,
  "isLoggedIn": false,
}
`;

exports[`/app/reducers/user should reduce a loadLocationsSuccess 1`] = `
Object {
  "_id": "u-1",
  "activeLocationId": "l-1",
  "isLoggedIn": true,
}
`;

exports[`/app/reducers/user should reduce a logout failure 1`] = `
Object {
  "isLoggedIn": false,
  "status": "failed",
}
`;

exports[`/app/reducers/user should reduce a logout request 1`] = `
Object {
  "isLoggedIn": false,
  "status": "logout",
}
`;

exports[`/app/reducers/user should reduce a logout success 1`] = `
Object {
  "isLoggedIn": false,
  "status": "logout",
}
`;

exports[`/app/reducers/user should reduce a missing location 1`] = `
Object {
  "_id": "u-1",
  "activeLocationId": false,
  "isLoggedIn": true,
}
`;
